// @vtex/clients/build/utils/search.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.searchEncodeURI = exports.SearchCrossSellingTypes = void 0;
const utils_1 = require("@gocommerce/utils");
// eslint-disable-next-line no-restricted-syntax
var SearchCrossSellingTypes;
(function (SearchCrossSellingTypes) {
    SearchCrossSellingTypes["whoboughtalsobought"] = "whoboughtalsobought";
    SearchCrossSellingTypes["similars"] = "similars";
    SearchCrossSellingTypes["whosawalsosaw"] = "whosawalsosaw";
    SearchCrossSellingTypes["whosawalsobought"] = "whosawalsobought";
    SearchCrossSellingTypes["accessories"] = "accessories";
    SearchCrossSellingTypes["suggestions"] = "suggestions";
})(SearchCrossSellingTypes = exports.SearchCrossSellingTypes || (exports.SearchCrossSellingTypes = {}));
exports.searchEncodeURI = (account) => (str) => {
    if (!utils_1.Functions.isGoCommerceAcc(account)) {
        return str.replace(/[%"'.()]/g, (c) => {
            switch (c) {
                case '%':
                    return '@perc@';
                case '"':
                    return '@quo@';
                case "'":
                    return '@squo@';
                case '.':
                    return '@dot@';
                case '(':
                    return '@lpar@';
                case ')':
                    return '@rpar@';
                default: {
                    return c;
                }
            }
        });
    }
    return str;
};
