// @gocommerce/utils/dist/logger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends4 = require('babel-runtime/helpers/extends');

var _extends5 = _interopRequireDefault(_extends4);

exports.default = function (ctx, request, response, TOKEN) {
    var vtex = ctx.vtex;
    var requestKeys = ['url', 'method', 'headers', 'body', 'data'];
    var responseKeys = ['status', 'statusText', 'message', 'headers', 'body', 'data'];
    var requestData = requestKeys.reduce(function (prev, next) {
        return (0, _extends5.default)({}, prev, (0, _defineProperty3.default)({}, 'request-' + next, request[next]));
    }, {});
    var responsetData = responseKeys.reduce(function (prev, next) {
        return (0, _extends5.default)({}, prev, (0, _defineProperty3.default)({}, 'response-' + next, response[next]));
    }, {});
    var flattterApiData = {};
    // const requestData = getSubset(requestKeys, request)
    // const responsetData = getSubset(responseKeys, response)
    // const flattterApiData = flatter({ vtex: vtex })
    //const rawMessage = objectToLogger(flattterApiData)
    var SOURCE = 'gc_ui';
    var headers = ctx.request.headers;
    var host = headers['x-forwarded-host'];
    var log = function log() {
        return {
            time: +new Date(),
            host: host,
            source: SOURCE,
            event: (0, _extends5.default)({}, flattterApiData, requestData, responsetData, { sourcetype: 'devLog' }),
            sourcetype: 'json',
            index: SOURCE
        };
    };
    return (0, _axios2.default)({
        url: SPLUNK_URI,
        method: 'POST',
        data: log(),
        headers: {
            'X-Vtex-Proxy-To': 'https://splunk-heavyforwarder-public.vtex.com:8088',
            'Proxy-Authorization': vtex.authToken,
            Authorization: 'Splunk ' + TOKEN,
            'Content-Type': 'application/json'
        }
    }).then(function (res) {
        // console.log(`detail log `)
    }).catch(function () {
        return console.log('detail log fail');
    }); //console.error)
};

var _v = require('uuid/v4');

var _v2 = _interopRequireDefault(_v);

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// import { flatter } from './functions'
var SOURCE = 'gc_ui';
var SPLUNK_URI = 'http://splunk-heavyforwarder-public.vtex.com/services/collector/event?channel=' + (0, _v2.default)() + '&source=' + SOURCE;
