// @vtex/clients/build/clients/affiliate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Affiliate = void 0;
const api_1 = require("@vtex/api");
const request_1 = require("../utils/request");
const baseURL = 'api/fulfillment/pvt/affiliates';
const routes = {
    affiliate: (id) => `${baseURL}/${id}`,
};
class Affiliate extends api_1.JanusClient {
    constructor(ctx, options) {
        super(ctx, {
            ...options,
        });
    }
    registerAffiliate({ name, id, salesChannelId, searchEndpoint, followUpEmail }, authMethod = 'AUTH_TOKEN', tracingConfig) {
        const metric = 'affiliate-registerAffiliate';
        return this.http.put(routes.affiliate(id), {
            followUpEmail,
            id,
            name,
            salesChannel: salesChannelId,
            searchURIEndPoint: searchEndpoint,
            searchURIEndPointAvailableVersions: ['1.x.x'],
            searchURIEndPointVersion: '1.x.x',
            useSellerPaymentMethod: false,
        }, request_1.getRequestConfig(this.context, authMethod, metric, tracingConfig));
    }
}
exports.Affiliate = Affiliate;
