// @vtex/clients/build/clients/omsApiProxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OMSProxy = void 0;
const api_1 = require("@vtex/api");
/**
 * Used to perform calls on OMS API in the VTEX infrastructure, to which you must declare an outbound policy.
 *
 * Policy:
 * {
 *   "name": "vtex.oms-api-proxy:oms-order-information"
 * },
 *
 */
class OMSProxy extends api_1.AppClient {
    constructor(context, options) {
        super('vtex.oms-api-proxy@0.x', context, options);
    }
    orders(id) {
        return this.http.get(`/orders/${id}`, {
            metric: 'orders-get',
        });
    }
    orderFormId(id) {
        return this.http.get(`orders/${id}/orderFormId`, {
            metric: 'orderFormId-get',
        });
    }
    customData(id, body) {
        return this.http.post(`orders/${id}/customData`, body, {
            metric: 'customData-post',
        });
    }
    register(body) {
        return this.http.post('register', body, {
            metric: 'registerCustomData-post',
        });
    }
}
exports.OMSProxy = OMSProxy;
